federatedgroups/lib/MixedGroupShareProvider.php

```diff
@@ -0,0 +1,184 @@
+<?php
+
+// Copyright (c) 2018, ownCloud GmbH
+// SPDX-FileCopyrightText: 2022 SURF
+//
+// SPDX-License-Identifier: AGPL-3.0-or-later
+
+namespace OCA\FederatedGroups;
+
+use OCA\FederatedGroups\FederatedFileSharing\Notifications;
+use OC\Share20\DefaultShareProvider;
+use OCP\Share\IShareProvider;
+use OCP\IDBConnection;
+use OCP\IUserManager;
+use OCP\IGroupManager;
+use OCP\Files\IRootFolder;
+
+
+/**
+ * Class MixedGroupShareProvider
+ *
+ * @package OC\Share20
+ */
+class MixedGroupShareProvider extends DefaultShareProvider implements IShareProvider {
+	// For representing foreign group members
+	// e.g. 'marie#oc2.docker'
+	public const SEPARATOR = '#';
+
+	/** @var Notifications */
+	private $notifications;
+
+	/**
+	 * Note $dbConn is private in the parent class
+	 * so we need to keep a copy of it here
+	 * @var IDBConnection
+	 */
+	private $dbConn;
+
+	/**
+	 * DefaultShareProvider constructor.
+	 *
+	 * @param IDBConnection $dbConn
+	 * @param IUserManager $userManager
+	 * @param IGroupManager $groupManager
+	 * @param IRootFolder $rootFolder
+	 * @param Notifications $notifications
+	 */
+	public function __construct(
+		IDBConnection $dbConn,
+		IUserManager $userManager,
+		IGroupManager $groupManager,
+		IRootFolder $rootFolder,
+		Notifications $notifications
+	) {
+		parent::__construct(
+			 $dbConn,
+			 $userManager,
+			 $groupManager,
+			 $rootFolder
+		);
+		error_log("Constructing the MixedGroupShareProvider");
+		// $this->notifications = $notifications;
+		$this->dbConn = $dbConn;
+	}
+
+	private function sendOcmInvite($getSharedBy, $shareOwner, $fedGroupId, $remote, $name) {
+		error_log("Send OCM invite ($getSharedBy, $shareOwner, $fedGroupId, $remote, $name)");
+	}
+
+	private function customGroupHasForeignersFrom($remote, $customGroupId) {
+		$queryBuilder = $this->dbConn->getQueryBuilder();
+		$cursor = $queryBuilder->select('user_id')->from('custom_group_member')
+			->where($queryBuilder->expr()->eq('group_id', $queryBuilder->createNamedParameter($customGroupId, IQueryBuilder::PARAM_INT)))
+			->where($queryBuilder->expr()->like('user_id', $queryBuilder->createNamedParameter("%#$remote", IQueryBuilder::PARAM_STR)))
```

Comment:
why it should check the remote? 
We should check the local users who are members of the given group 
OR
the entries are like this phrase `%#$groupId@%`

Reply:
If `marie#oc2.docker` was already a member of the group and `john#oc2.docker` is added, there is no need to send an OCM share to `federalists@oc2.docker`.
But if `john#oc3.docker` is added and there are no members `LIKE "%#oc3.docker"` yet, then it's necessary to send an OCM share to `federalists@oc3.docker` regarding this share.